### DISCUSSION TITLE
backend.ai run --rm --exec 'python single_mnist.py' tf2 ./single_mnist.py ./mnist.py 실행시에  그냥 killed 됩니다.

### DISCUSSION THREAD

**Discussion by eunjungchoi** [lablup/contribution-academy-2021#83] [Q&A]
`$ backend.ai run --rm --exec 'python single_mnist.py' tf2 ./single_mnist.py ./mnist.py ` 명령어를 실행하면 
 
`2021-08-22 11:32:06.634942: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0` 까지는 잘 진행되다가 

갑자기 중간에 killed 됩니다. 그리고 그냥 세션이 종료되어 버려요. 

혹시 이런 문제를 겪은 분들이 계실까요? 

<img width="1638" alt="스크린샷 2021-08-22 오후 8 34 06" src="https://user-images.githubusercontent.com/5889435/130353557-b27836bf-88e4-4a55-be62-930916714af0.png">

이 때  agent server의 상태는 이렇습니다. 뭔가 제대로 execution이 안 된 상태로 그대로 destroy kernel 되버리네요.  딱히 에러 메시지를 뱉고 있지도 않습니다 
<img width="1640" alt="스크린샷 2021-08-22 오후 8 36 28" src="https://user-images.githubusercontent.com/5889435/130353634-16dd6606-eec0-4096-a7e8-7c2c95eab704.png">

이유를 모르겠네요.  
killed 되기 전 중간 과정에서  backend.ai ps 로 체크해보면  다음과 같이 떴구요. 
<img width="1646" alt="스크린샷 2021-08-22 오후 8 38 49" src="https://user-images.githubusercontent.com/5889435/130353701-96001a42-bda2-44b8-bc1f-42dae5a2ff3f.png">

클라이언트에서 체크해보면 이렇습니다. 
<img width="1568" alt="스크린샷 2021-08-22 오후 8 57 52" src="https://user-images.githubusercontent.com/5889435/130354251-de1c1846-613c-4336-8354-22d4111d1226.png">

일단 tf2 이미지는  docker pull 로 받아왔었습니다.  
`docker pull cr.backend.ai/stable/python-tensorflow:2.4-py38-cuda11.1 `

그리고 killed로 지금은 바로 끝나버리고 있는데  최초에 구동했을 때는 kernel_id를 못 찾는다는 runtime error가 agent server에서 발생했었습니다. 

<img width="1640" alt="kernel_id_failed" src="https://user-images.githubusercontent.com/5889435/130353888-60c5d6de-540f-45a3-b835-4c496359663b.png">

아무런 조치를 취하지 않고, 다시 같은 명령어를 구동했을 때 이제 그냥 killed 되어 버리고 있구요. 

+)
클라이언트에서  세션 > 종료 >   종료된 세션의 '제어' 버튼을 눌러서 log를 확인해보면 이렇습니다. 

<img width="1161" alt="스크린샷 2021-08-22 오후 9 05 13" src="https://user-images.githubusercontent.com/5889435/130354457-f4ac5a18-032b-4972-b1eb-58a246b6a9a5.png">

**Comment by eunjungchoi** [comment]
찾아보니 memory와 관련이 있었네요. 
tensorflow의 메모리를 1G로 다시 설정해주니  (다른 것들도 기본이 메모리 1G)  동작합니다. 

<img width="1528" alt="스크린샷 2021-08-22 오후 9 15 56" src="https://user-images.githubusercontent.com/5889435/130354744-bd5788d2-0c00-4803-b76c-2da6aacea1ef.png">


다만 이제 kill 되지 않고 동작하는데  모듈 내에서 tf 를 못찾는 에러가 발생하네요. 
<img width="1554" alt="스크린샷 2021-08-22 오후 9 17 34" src="https://user-images.githubusercontent.com/5889435/130354780-e7f47617-50dd-410c-ab0d-1c7833c117fc.png">

**Reply by inureyes** [reply]
@eunjungchoi  음 single_mnist.py의 내용을 여기 붙여주실 수 있을런지요?

**Reply by eunjungchoi** [reply]
mnist.py  파일 내에  최상단에 

`import tensorflow as tf`

`import numpy as np` 

이 두 줄을 추가해주니 바로 해결되었습니다. 😅   이 부분이 빠져있었나봐요. 

감사합니다.

**Reply by eunjungchoi** [reply]
잘 실행되었습니다.  휴.  감사합니다. 
 
<img width="1225" alt="스크린샷 2021-08-22 오후 9 32 41" src="https://user-images.githubusercontent.com/5889435/130355303-c05c78cf-3899-4800-8c9c-cc3df76a71fa.png">

**Reply by inureyes** [reply]
수고하셨습니다!
